cpp/modmesh/python/common.cpp

```diff
@@ -171,7 +171,52 @@ void Interpreter::exec_code(std::string const & code)
 {
     // NOLINTNEXTLINE(misc-const-correctness)
     pybind11::object mod_sys = pybind11::module_::import("modmesh.system");
-    mod_sys.attr("exec_code")(code);
+    try
+    {
+        mod_sys.attr("exec_code")(code);
+    }
+    catch (const pybind11::error_already_set & e)
+    {
+        std::cerr << e.what() << std::endl;
+    }
+}
+
+PyStdErrOutStreamRedirect::PyStdErrOutStreamRedirect()
+{
+    auto sys_module = pybind11::module::import("sys");
+    m_stdout = sys_module.attr("stdout");
+    m_stderr = sys_module.attr("stderr");
+    auto string_io = pybind11::module::import("io").attr("StringIO");
+    m_stdout_buffer = string_io(); // Other file like object can be used here as well, such as objects created by pybind11
+    m_stderr_buffer = string_io();
+    sys_module.attr("stdout") = m_stdout_buffer;
+    sys_module.attr("stderr") = m_stderr_buffer;
+}
+
+std::string PyStdErrOutStreamRedirect::stdout_string()
+{
+    m_stdout_buffer.attr("seek")(0);
+    return pybind11::str(m_stdout_buffer.attr("read")());
+}
+
+std::string PyStdErrOutStreamRedirect::stderr_string()
+{
+    m_stderr_buffer.attr("seek")(0);
+    return pybind11::str(m_stderr_buffer.attr("read")());
+}
+
+PyStdErrOutStreamRedirect::~PyStdErrOutStreamRedirect() noexcept(true)
```

Comment:
Because you still may throw an exception, it should be `noexcept(false)`.

Reply:
because I only catch `pybind11::error_already_set` here?

also, if I use `try  catch(...)` for the whole deconstruction function, will it throw?

Reply:
I do not fully understand your questions.  Could you be more specific?

Reply:
> Because you still may throw an exception, it should be `noexcept(false)`

is it because I only catch `pybind11::error_already_set` in the code?

if I write:
```
    try
    {
     // ..
    }
    catch (...) // capture all exception
    {
        // ...
    }
```

will it still throw exception?

Reply:
I do not know.  `noexcept` means the function does not throw exception.  I am not sure it means no exception handling is allowed in the function, or no except is thrown outside the function.  I guess it's the latter, but if it's the former, a `try catch` block in `noexcept` would be problematic.

I would be interested in the behavior too.

Reply:
it seems no difference for the behavior when `noexcept` is `true` or `false`

```
struct Foo {
    Foo(){}
    
    ~Foo() noexcept(true){
        try {
            throw "123";
        }
        catch(...) {
            //
        }
    }
};

int main()
{
  Foo foo;
}
```

the program can run without panic with both conditions.

Reply:
however, this panic
```
struct Foo {
    Foo(){}

    ~Foo() noexcept(true){
        try {
            throw "123";
        }
        catch(int a) {
                std::cerr << "XX" ;
        }
    }
};

int main()
{
  Foo foo;
}
```

Reply:
![image](https://user-images.githubusercontent.com/33109379/203799686-1ed7d207-d436-43a5-976c-2938d1c39b27.png)
https://en.cppreference.com/w/cpp/language/noexcept_spec

Reply:
so I think the answer is, the current code might need to use `noexcept(false)` because it only capture `error_already_set`.
if it is `catch(...)` then we can set it `noexcept(true)`